docs/chymyst07.md

```diff
@@ -0,0 +1,731 @@
+What we would like to allow is _any_ code anywhere to contain any number of critical sections identified by the same `beginCritical()` call.
+
+How can we implement this functionality using the chemical machine?
+Since the only way to communicate between reactions is to emit molecules, `beginCritical()` and `endCritical()` must be molecules.
+Clearly, `beginCritical` must be blocking while `endCritical` does not have to be; so let us make `endCritical` a non-blocking molecule.
+
+What should happen when a process emits `beginCritical()`?
+We must enforce a single-process access to the critical section.
+In other words, a reaction that consumes `beginCritical()` should only start one at a time even if several `beginCritical` molecules are available.
+Therefore, this reaction must also require another input molecule, say `access()`, of which we will only have a single copy emitted into the soup:
+
+```scala
+val beginCritical = b[Unit, Unit]
+val access = m[Unit]
+
+site(
+  go { case beginCritical(_, reply) + access(_) => ???; reply(); ??? }
+)
+access() // Emit only one copy.
+
+```
+
+Just as in the case of single-access resource, we have guaranteed that `beginCritical()` blocks if called more than once.
+All we need to do is insure that there is initially a single copy of `access()` in the soup, and that no further copies of `access()` are ever emitted.
+
+Another requirement is that emitting `endCritical()` must end whatever `beginCritical()` began, but only if `endCritical()` is emitted by the _same reaction_.
+If a different reaction emits `endCritical()`, access to the critical section should not be granted.
+Somehow, the `endCritical()` molecules must be different when emitted by different reactions (however, `beginCritical()` must be the same for all reactions, or else there can't be any contention between them).
+
+Additionally, we would like it to be impossible for any reaction to call `endCritical()` without first calling `beginCritical()`.
+
+One way of achieving this is to make `beginCritical()` return a value that enables us to call `endCritical()`.
+For instance, `beginCritical()` could actually create a new, unique emitter for `endCritical()`, and return that emitter:
+
+```scala
+val beginCritical = b[Unit, M[Unit]]
+val access = m[Unit]
+
+site (
+  go { case beginCritical(_, reply) + access(_) =>
+    val endCritical = m[Unit] // Declare a new emitter.
+    site (
+      go { case endCritical(_) => ??? }
+    )
+    reply(endCritical) // beginCritical() returns the new emitter.
+  }
+)
+access() // Emit only one copy.
+
+```
+
+Since `endCritical` and its reaction site are defined within the local scope of the reaction that consumes `beginCritical()`, each such reaction will create a _chemically unique_ new molecule that no other reactions can emit.
```

Comment:
A lot to digest when reading a bit fast, but this is very impressive. You must have known long ago that this should be possible... Maybe the previous projects or papers made claims that critical sections could be implemented with JC.